packages/mermaid/src/rendering-util/handle-markdown-text.ts

```diff
@@ -0,0 +1,53 @@
+import { micromark } from 'micromark';
+import { fromMarkdown } from 'mdast-util-from-markdown';
+import type { Content } from 'mdast';
+
+function preprocessMarkdown(markdown: string): string {
+  // Replace multiple newlines with a single newline
+  const withoutMultipleNewlines = markdown.replace(/\n{2,}/g, '\n');
+  // Remove extra spaces at the beginning of each line
+  const withoutExtraSpaces = withoutMultipleNewlines.replace(/^\s+/gm, '');
+  return withoutExtraSpaces;
+}
+
+export function markdownToLines(markdown: string) {
+  const preprocessedMarkdown = preprocessMarkdown(markdown);
+  const { children } = fromMarkdown(preprocessedMarkdown);
+  const lines: { content: string; type: string }[][] = [[]];
+  let currentLine = 0;
+
+  function processNode(node: Content, parentType?: string) {
+    if (node.type === 'text') {
+      const textLines = node.value.split('\n');
+      textLines.forEach((textLine, index) => {
+        if (index !== 0) {
+          currentLine++;
+          lines.push([]);
+        }
+        textLine.split(' ').forEach((word) => {
+          if (word) {
+            lines[currentLine].push({ content: word, type: parentType || 'normal' });
+          }
+        });
+      });
+    } else if (node.type === 'strong' || node.type === 'emphasis') {
+      node.children.forEach((contentNode) => {
+        processNode(contentNode, node.type);
+      });
+    }
+  }
+
+  children.forEach((treeNode) => {
+    if (treeNode.type === 'paragraph') {
+      treeNode.children.forEach((contentNode) => {
+        processNode(contentNode);
+      });
+    }
+  });
+
+  return lines;
+}
+
+export function markdownToHTML(markdown: string): string {
+  return micromark(preprocessMarkdown(markdown)).replaceAll('\n', '<br/>');
```

Comment:
@remcohaszing I didn't use `[remark-breaks](https://github.com/remarkjs/remark-breaks)` to add the `<br>` tags as using remark would need the code to be async, which might need larger changes in the codebase. 

Does this have any issue I might've missed?

Reply:
I recommend against treating newlines as linebreaks in markdown, as they should be ignored according to commonmark.

If you really do want to, you can though. `remark-break` isn’t async. Unified processors have a `processSync` function alongside `process` that also works if all plugins are synchronous.

You could also copy the [logic](https://github.com/remarkjs/remark-breaks/blob/main/index.js#L18-L47) from `remark-breaks`. It’s not ideal, but it’s not that much either.

Reply:
My guess is that the line above is `.use(remarkRehype)`

Try:

```ts
import * as hast from 'hast'
import * as mdast from 'mdast'
import { Plugin } from 'unified'

// …
.use(remarkRehype as Plugin<[], mdast.Root, hast.Root>)
```

Reply:
It's not just a TS error. The output is an AST(?), not HTML. I tried first with a `@ts-ignore`

<img width="390" alt="image" src="https://user-images.githubusercontent.com/10703445/234371195-ade715ed-7efd-4b01-a77a-098a42e66dde.png">

This is the code.

```ts
return unified()
    .use(remarkParse)
    .use(remarkBreaks)
    .use(remarkRehype)
    .use(rehypeSanitize)
    .use(rehypeStringify)
    .parseSync(markdown);
```

Reply:
- [`parse`](https://github.com/unifiedjs/unified#processorparsefile) parses the input to AST.
- [`run`](https://github.com/unifiedjs/unified#processorruntree-file-done) / [`runSync`](https://github.com/unifiedjs/unified#processorrunsynctree-file) transforms the AST.
- [`stringify`](https://github.com/unifiedjs/unified#processorstringifytree-file) compiles the AST.
- `process` / `processSync` does `parse` → `run` → `compile`